src/coreclr/src/jit/morph.cpp

```diff
@@ -11611,6 +11611,68 @@ GenTree* Compiler::fgMorphFieldAssignToSIMDIntrinsicSet(GenTree* tree)
 
 #endif // FEATURE_SIMD
 
+//------------------------------------------------------------------------------
+// fgMorphCommutative : Try to simplify "(X op C1) op C2" to "X op (C1 op C2)"
+//                      for commutative operators.
+//
+// Arguments:
+//       tree - node to fold
+//
+// return value:
+//       A folded GenTree* instance or nullptr if something prevents folding.
+//
+
+GenTree* Compiler::fgMorphCommutative(GenTreeOp* tree)
+{
+    assert(varTypeIsIntegralOrI(tree->TypeGet()));
+    assert(tree->OperIs(GT_ADD, GT_MUL, GT_OR, GT_AND, GT_XOR));
+
+    // op1 can be GT_COMMA, in this case we're going to fold
+    // "COMMA(..., (X op C1)) op C2" to "COMMA(..., X op (C1 op C2))"
+    GenTree*   op1  = tree->gtGetOp1()->gtEffectiveVal(true);
+    genTreeOps oper = tree->OperGet();
+
+    if (!op1->OperIs(oper) || !tree->gtGetOp2()->IsCnsIntOrI() || !op1->gtGetOp2()->IsCnsIntOrI() ||
```

Comment:
Do we call this before or after `fgMorphSmpOpOptional` has moved constants to the right? Seems like it happens before, so maybe we should call `fgMorphCommutative ` from `fgMorphSmpOpOptional` after it calls `fgMoveOpsLeft`?

(perhaps try an example like `2 + x + 2` ?)

Reply:
Looks like that rotation happens before `fgMorphCommutative`:
![image](https://user-images.githubusercontent.com/523221/96767775-a3b63d00-13e5-11eb-839d-99cde2b4ccc7.png)